keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,537 @@
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *PodContainerSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // If specified, controller only select the containers that are listed from the selected pods based on PodSelector. 
+	// Otherwise, all containers of the pods will be selected
+        // +optional
+        ContainerList []string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        HookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerName string
+
+        // If not set, it is nil, indicating Action has not started
+        // If set, it means Action has started at the specified time
+	// +optional
+        Timestamp *int64
+
+        // ActionSucceed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after ActionTimeoutSeconds passes.
+        // +optional
+        Succeed *bool
+
+        // The last error encountered when executing the action. The hook controller might update this field each time
+	// it retries the execution.
+	// +optional
+        Error *HookError
+}
+
+type HookError struct {
+        // Type of the error
+        // This is required
+        ErrorType ErrorType
+
+        // Error message
+        // +optional
+        Message *string
```

Comment:
What is the distinction between Message and Reason ?

Reply:
"Message" is a brief description of the error.  "Reason" can provide more details on why it failed.  Maybe we can combine the two and just have "Message".